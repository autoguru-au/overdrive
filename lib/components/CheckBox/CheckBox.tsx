import { CheckIcon } from '@autoguru/icons';
import clsx from 'clsx';
import * as React from 'react';
import { FunctionComponent, memo, ReactNode } from 'react';
import { useStyles } from 'react-treat';

import { CheckableBase } from '../CheckableBase';
import { useCheckableStyles } from '../CheckableBase/useCheckableStyles';
import { Icon } from '../Icon';
import * as styleRefs from './CheckBox.treat';

interface Props {
	className?: string;
	checked?: boolean;
	disabled?: boolean;
	name?: string;
	value: string;
	children: ReactNode;

	onClick?(checked: boolean): void;

	onChange?(checked: boolean): void;
}

export const CheckBox: FunctionComponent<Props> = memo(
	({
		value,
		className = '',
		name = '',
		disabled = false,
		checked = false,
		onClick = () => void 0,
		onChange = () => void 0,
		children,
	}) => {
		const styles = useStyles(styleRefs);
		const { checkable, checkableItem } = useCheckableStyles();

		return (
			<CheckableBase
				inputType="checkbox"
				className={clsx([styles.checkbox, className])}
				inputName={name}
				value={value}
				label={children}
				disabled={disabled}
				checked={checked}
				handleClick={onClick}
				handleChange={onChange}>
				<div className={checkable}>
					{checked && (
						<Icon
							className={styles.icon}
							icon={CheckIcon}
							size="small"
						/>
					)}
					<div
						className={clsx(checkableItem, styles.base.default, {
							[styles.base.selected]: checked,
						})}
					/>
				</div>
			</CheckableBase>
		);
	},
);
