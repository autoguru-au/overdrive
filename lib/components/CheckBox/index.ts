export { CheckBox, Props as ICheckBoxProps } from './CheckBox';
