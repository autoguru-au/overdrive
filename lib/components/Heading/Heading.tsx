import React from 'react';

import type { Sprinkles } from '../../styles/sprinkles.css';
import { Box, type BoxProps } from '../Box';
import { textStyles, type TextStylesProps } from '../Text/textStyles';

export type HeadingTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';

export interface HeadingProps
	extends Omit<BoxProps, keyof TextStylesProps>,
		Omit<TextStylesProps, 'as'> {
	as?: HeadingTags;
	align?: TextStylesProps['textAlign'];
	/** @deprecated Prefer `as` */
	is?: HeadingTags;
}

const defaultSizeMap: Record<HeadingTags, Sprinkles['fontSize']> = {
	h1: '8',
	h2: '7',
	h3: '6',
	h4: '4',
	h5: '3',
	h6: '2',
};

/**
 * Heading renders an <h1... h6> with default font size and weight
 */
export const Heading = ({
	is = 'h1',
	as = is,
	align,
	breakWord,
	children,
	className,
	color, // semantic tokens
	colour = color ? undefined : 'dark', // legacy intentional tokens
	fontWeight = 'bold',
	noWrap,
	size = defaultSizeMap[as],
	transform,
	...props
}: HeadingProps) => (
	<Box
		{...props}
		as={as}
		color={color}
		className={[
			textStyles({
				breakWord,
				colour,
				fontWeight,
				noWrap,
				size,
				textAlign: align,
				transform,
			}),
			className,
		]}
	>
		{children}
	</Box>
);
