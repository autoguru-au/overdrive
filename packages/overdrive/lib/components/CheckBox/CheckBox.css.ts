import { style, styleVariants } from '@vanilla-extract/css';

import { themeContractVars as vars } from '../../themes/theme.css';

export const icon = style({
	zIndex: 1,
	top: `calc(50% - 0.5*(${vars.space['4']}))`,
	left: `calc(50% - 0.5*(${vars.space['4']}))`,
	width: vars.space['4'],
	height: vars.space['4'],
});

export const base = styleVariants({
	default: {
		zIndex: 0,
		borderRadius: vars.border.radius['min'],
		top: vars.space['3'],
		left: vars.space['3'],
		width: vars.space['4'],
		height: vars.space['4'],
		transition: `border-color 0.2s ${vars.animation.easing.decelerate} 0s, background-color 0.2s ${vars.animation.easing.decelerate} 0s`,
	},

	selected: {
		borderColor: vars.colours.intent.primary.background.standard,
		backgroundColor: vars.colours.intent.primary.background.standard,
	},
});
