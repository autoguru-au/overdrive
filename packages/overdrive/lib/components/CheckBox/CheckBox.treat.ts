import { style, styleMap } from 'treat';

export const icon = style((theme) => ({
	zIndex: 1,
	top: `calc(50% - 0.5*(${theme.space['4']}))`,
	left: `calc(50% - 0.5*(${theme.space['4']}))`,
	width: theme.space['4'],
	height: theme.space['4'],
}));

export const base = styleMap(({ space, animation, colours }) => ({
	default: {
		zIndex: 0,
		borderRadius: '2px',
		top: space['3'],
		left: space['3'],
		width: space['4'],
		height: space['4'],
		transition: `border-color 0.2s ${animation.easing.decelerate} 0s, background-color 0.2s ${animation.easing.decelerate} 0s`,
	},

	selected: {
		borderColor: colours.intent.primary.background,
	},
}));
