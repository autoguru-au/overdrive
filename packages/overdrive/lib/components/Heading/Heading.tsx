import * as React from 'react';
import { ComponentProps, FunctionComponent } from 'react';
import type { Theme } from 'treat/theme';

import { Box } from '../Box';
import type { TextStyleProps } from '../Text';
import { useTextStyles } from '../Text';

export interface Props
	extends Omit<TextStyleProps, 'is'>,
		Pick<ComponentProps<typeof Box>, 'id'> {
	className?: string;
	is?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
	colour?: Exclude<keyof Theme['typography']['colour'], 'muted'>;
}

const sizeScaleDefaults = {
	h1: '8',
	h2: '7',
	h3: '6',
	h4: '4',
	h5: '3',
	h6: '2',
} as const;

export const Heading: FunctionComponent<Props> = ({
	is = 'h1',
	align,
	fontWeight = 'bold',
	noWrap,
	transform,
	colour = 'dark',
	size = sizeScaleDefaults[is],
	id,
	className = '',
	children,
}) => (
	<Box
		id={id}
		is={is}
		className={[
			useTextStyles({
				size,
				align,
				colour,
				noWrap,
				transform,
				fontWeight,
			}),
			className,
		]}>
		{children}
	</Box>
);
