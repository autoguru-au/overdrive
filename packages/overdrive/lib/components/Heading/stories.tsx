import { select } from '@storybook/addon-knobs';
import * as React from 'react';

import { Heading } from '.';

const sizeScale = ['1', '2', '3', '4', '5', '6', '7', '8', '9'];
const alignOptions: ['left', 'center', 'right'] = ['left', 'center', 'right'];
const colourOptions = [
	'dark',
	'white',
	'light',
	'danger',
	'warning',
	'success',
	'information',
];

const baseProps = () => ({
	size: select('Size', sizeScale, '4'),
	align: select('align', alignOptions, 'left'),
	colour: select('Colour', colourOptions, 'dark'),
});

export default {
	title: 'Foundation|Typography/Heading',
	component: Heading,
	decorators: [
		(story) => (
			<div style={{ maxWidth: '350px', width: '100%' }}>{story()}</div>
		),
	],
};

export const standard = () => (
	<Heading
		is={select('Tag', ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'], 'h1')}
		{...baseProps()}>
		Hello World
	</Heading>
);

export const allSizes = () => (
	<div>
		<Heading is="h6">I am a heading</Heading>
		<Heading is="h5">I am a heading</Heading>
		<Heading is="h4">I am a heading</Heading>
		<Heading is="h3">I am a heading</Heading>
		<Heading is="h2">I am a heading</Heading>
		<Heading is="h1">I am a heading</Heading>
	</div>
);
