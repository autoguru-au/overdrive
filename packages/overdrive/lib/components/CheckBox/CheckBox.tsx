import { CheckIcon } from '@autoguru/icons';
import clsx from 'clsx';
import * as React from 'react';
import { forwardRef, ReactNode } from 'react';
import { useStyles } from 'react-treat';

import { noop } from '../../utils';
import { Box, useBoxStyles } from '../Box';
import { Icon } from '../Icon';
import { useTextStyles } from '../Text';
import { CheckableBase } from '../private/CheckableBase';
import { useCheckableStyles } from '../private/CheckableBase/useCheckableStyles';

import * as styleRefs from './CheckBox.treat';

export interface Props {
	className?: string;
	checked?: boolean;
	disabled?: boolean;
	name?: string;
	value: string;
	children: ReactNode;

	onClick?(checked: boolean): void;

	onChange?(checked: boolean): void;
}

export const CheckBox = forwardRef<HTMLInputElement, Props>(
	(
		{
			value,
			className = '',
			name = '',
			disabled = false,
			checked = false,
			onClick = noop,
			onChange = noop,
			children,
		},
		ref,
	) => {
		const styles = useStyles(styleRefs);
		const iconStyles = clsx(
			useTextStyles({ colour: 'white' }),
			useBoxStyles({ position: 'absolute' }),
		);
		const { checkableItem } = useCheckableStyles();

		return (
			<CheckableBase
				ref={ref}
				inputType="checkbox"
				className={className}
				inputName={name}
				value={value}
				label={children}
				disabled={disabled}
				checked={checked}
				handleClick={onClick}
				handleChange={onChange}>
				{checked ? (
					<Icon
						className={clsx(styles.icon, iconStyles)}
						icon={CheckIcon}
						size="small"
					/>
				) : null}
				<Box
					borderWidth="2"
					borderColour="gray"
					backgroundColour={checked ? 'green600' : 'transparent'}
					className={clsx(checkableItem, styles.base.default, {
						[styles.base.selected]: checked,
					})}
				/>
			</CheckableBase>
		);
	},
);
