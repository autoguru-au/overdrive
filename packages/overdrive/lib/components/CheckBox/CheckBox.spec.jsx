import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';

import { Heading } from '../Heading';
import { CheckBox } from './CheckBox';

describe('<CheckBox />', () => {
	it('should not throw', () => {
		expect(() => render(<CheckBox />)).not.toThrow();
	});

	it('should match the snapshot for a single check with no value or label', () => {
		expect(render(<CheckBox />).container.firstChild).toMatchSnapshot();
	});

	it('should match the snapshot for a single check', () => {
		expect(
			render(<CheckBox children="check label 1" value="1" />).container
				.firstChild,
		).toMatchSnapshot();
	});

	it('should match the snapshot for a checked check', () => {
		expect(
			render(<CheckBox children="check label 1" checked value="1" />)
				.container.firstChild,
		).toMatchSnapshot();
	});

	it('should match the snapshot for a disable checkbox', () => {
		expect(
			render(<CheckBox disabled children="check label 1" value="1" />)
				.container.firstChild,
		).toMatchSnapshot();
	});

	it.skip('should pass on className to dom element', () => {
		expect(
			render(<CheckBox className="check-class" />).container.firstChild,
		).toHaveClass('check-class');
	});

	it('should call the onClick function passed down to it when clicked', () => {
		const spyedClickCallback = jest.fn();

		const { container } = render(
			<CheckBox
				children="check label 1"
				value="1"
				onClick={spyedClickCallback}
			/>,
		);

		fireEvent.click(container.querySelector('input'));

		expect(spyedClickCallback).toHaveBeenCalledTimes(1);
	});

	it('should pass the checked value to the native element', () => {
		const { container } = render(
			<CheckBox children="check label 1" checked value="1" />,
		);

		expect(container.querySelector('input')).toHaveAttribute('checked', '');
	});

	it('should not throw is onChange callback is not attached', () => {
		const { container } = render(
			<CheckBox children="check label 1" value="1" />,
		);

		expect(() => {
			fireEvent.change(container.querySelector('input'), {});
		}).not.toThrow();
	});

	it('should call the onChange function passed down to it when checked value has changes', () => {
		const spyedChangeCallback = jest.fn();
		const { container } = render(
			<CheckBox
				children="check label 1"
				value="1"
				onChange={spyedChangeCallback}
			/>,
		);

		expect(() => {
			fireEvent.click(container.querySelector('input'), {});
		}).not.toThrow();

		expect(spyedChangeCallback).toHaveBeenCalledTimes(1);
	});

	describe('when with component child', () => {
		it('should match the snapshot', () => {
			expect(
				render(
					<CheckBox children="check label 1" value="1">
						<Heading>Hello checkbox</Heading>
					</CheckBox>,
				).container.firstChild,
			).toMatchSnapshot();
		});
	});
});
