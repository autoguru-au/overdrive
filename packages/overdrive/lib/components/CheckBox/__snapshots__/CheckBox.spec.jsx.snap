// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckBox /> should match the snapshot for a checked check 1`] = `
<div
  class="base 3_mobile_base 3_mobile_base 3_mobile_base none_mobile_base display_flex position_relative center_mobile_base row_mobile_base flexStart_mobile_base root_base base appearance cursorPointer button"
>
  <input
    checked=""
    class="base block appearance input none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base width_full height_full position_absolute base appearance cursorPointer button nativeInput_default nativeInput_checked"
    name=""
    type="checkbox"
    value="1"
  />
  <div
    class="base display_flex position_relative center_mobile_base center_mobile_base checkable_base base appearance cursorPointer button"
  >
    <i
      class="base display_block small_mobile_base icon_base root_base colours_white_base position_absolute"
      role="presentation"
    >
      <svg
        class="base display_block width_full height_full"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M21 7L9 19l-5.5-5.5 1.414-1.414L9 16.172 19.586 5.586 21 7z"
        />
      </svg>
    </i>
    <div
      class="base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base 2_mobile_base 2_mobile_base 2_mobile_base 2_mobile_base backgroundColours_green600_base checkableItem_base base_default_base base_selected_base"
    />
  </div>
  <label
    class="base width_full base appearance cursorPointer button root_base colours_neutral_base sizes_4_base"
  >
    <span
      class="base textAlign_left root_base colours_neutral_base fontWeight_normal_base sizes_4_base"
    >
      check label 1
    </span>
  </label>
</div>
`;

exports[`<CheckBox /> should match the snapshot for a disable checkbox 1`] = `
<div
  class="base 3_mobile_base 3_mobile_base 3_mobile_base none_mobile_base display_flex position_relative center_mobile_base row_mobile_base flexStart_mobile_base root_base base appearance cursorPointer button disabled"
>
  <input
    class="base block appearance input none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base width_full height_full position_absolute pointerEvents_none base appearance cursorPointer button nativeInput_default"
    disabled=""
    name=""
    type="checkbox"
    value="1"
  />
  <div
    class="base display_flex position_relative center_mobile_base center_mobile_base checkable_base base appearance cursorPointer button"
  >
    <div
      class="base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base 2_mobile_base 2_mobile_base 2_mobile_base 2_mobile_base backgroundColours_transparent_base checkableItem_base base_default_base"
    />
  </div>
  <label
    class="base width_full pointerEvents_none base appearance cursorPointer button root_base colours_neutral_base sizes_4_base label_disabled_base"
  >
    <span
      class="base textAlign_left root_base colours_neutral_base fontWeight_normal_base sizes_4_base"
    >
      check label 1
    </span>
  </label>
</div>
`;

exports[`<CheckBox /> should match the snapshot for a single check 1`] = `
<div
  class="base 3_mobile_base 3_mobile_base 3_mobile_base none_mobile_base display_flex position_relative center_mobile_base row_mobile_base flexStart_mobile_base root_base base appearance cursorPointer button"
>
  <input
    class="base block appearance input none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base width_full height_full position_absolute base appearance cursorPointer button nativeInput_default"
    name=""
    type="checkbox"
    value="1"
  />
  <div
    class="base display_flex position_relative center_mobile_base center_mobile_base checkable_base base appearance cursorPointer button"
  >
    <div
      class="base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base 2_mobile_base 2_mobile_base 2_mobile_base 2_mobile_base backgroundColours_transparent_base checkableItem_base base_default_base"
    />
  </div>
  <label
    class="base width_full base appearance cursorPointer button root_base colours_neutral_base sizes_4_base"
  >
    <span
      class="base textAlign_left root_base colours_neutral_base fontWeight_normal_base sizes_4_base"
    >
      check label 1
    </span>
  </label>
</div>
`;

exports[`<CheckBox /> should match the snapshot for a single check with no value or label 1`] = `
<div
  class="base 3_mobile_base 3_mobile_base 3_mobile_base none_mobile_base display_flex position_relative center_mobile_base row_mobile_base flexStart_mobile_base root_base base appearance cursorPointer button"
>
  <input
    class="base block appearance input none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base width_full height_full position_absolute base appearance cursorPointer button nativeInput_default"
    name=""
    type="checkbox"
    value=""
  />
  <div
    class="base display_flex position_relative center_mobile_base center_mobile_base checkable_base base appearance cursorPointer button"
  >
    <div
      class="base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base 2_mobile_base 2_mobile_base 2_mobile_base 2_mobile_base backgroundColours_transparent_base checkableItem_base base_default_base"
    />
  </div>
  <label
    class="base width_full base appearance cursorPointer button root_base colours_neutral_base sizes_4_base"
  >
    <span
      class="base textAlign_left root_base colours_neutral_base fontWeight_normal_base sizes_4_base"
    />
  </label>
</div>
`;

exports[`<CheckBox /> when with component child should match the snapshot 1`] = `
<div
  class="base 3_mobile_base 3_mobile_base 3_mobile_base none_mobile_base display_flex position_relative center_mobile_base row_mobile_base flexStart_mobile_base root_base base appearance cursorPointer button"
>
  <input
    class="base block appearance input none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base none_mobile_base width_full height_full position_absolute base appearance cursorPointer button nativeInput_default"
    name=""
    type="checkbox"
    value="1"
  />
  <div
    class="base display_flex position_relative center_mobile_base center_mobile_base checkable_base base appearance cursorPointer button"
  >
    <div
      class="base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base 2_mobile_base 2_mobile_base 2_mobile_base 2_mobile_base backgroundColours_transparent_base checkableItem_base base_default_base"
    />
  </div>
  <label
    class="base width_full base appearance cursorPointer button root_base colours_neutral_base sizes_4_base"
  >
    <h1
      class="base root_base colours_dark_base fontWeight_bold_base sizes_8_base"
    >
      Hello checkbox
    </h1>
  </label>
</div>
`;
